File: minSpantree.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strconv"
	"sort"
)

type Vertex struct {
	value int
	parent *Vertex
}

func (v *Vertex) getRoot() *Vertex {
	if v.parent == nil {
		return v;
	} else {
		return v.parent.getRoot()
	}
}

type Edge struct {
	u, v *Vertex
	weight int
}

type Edges []Edge

// Implement the sort.Interface for Edges
func (edges Edges) Len() int { return len(edges) }
func (edges Edges) Less(i, j int) bool { return edges[i].weight < edges[j].weight }
func (edges Edges) Swap(i, j int) { edges[i], edges[j] = edges[j], edges[i] }

func getNextInt(scanner *bufio.Scanner) int {
	scanner.Scan()
	i, _ := strconv.Atoi(scanner.Text());
	return i;
}

func main() {

	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	for test := getNextInt(scanner); test > 0; test-- {

		numVertices := getNextInt(scanner)
		numEdges := getNextInt(scanner)

		var vertices = make([]Vertex, numVertices)
		var edges Edges

		for i := 0; i < numEdges; i++ {

			uValue := getNextInt(scanner) - 1
			vValue := getNextInt(scanner) - 1
			vertices[uValue].value = uValue
			vertices[vValue].value = vValue

			edges = append(edges, Edge{ &vertices[uValue], &vertices[vValue], getNextInt(scanner) })

		}

		sort.Sort(edges)

		totalWeight := 0

		for _, edge := range edges {
			if edge.u.getRoot() != edge.v.getRoot() {
				edge.v.getRoot().parent = edge.u.getRoot();
				totalWeight += edge.weight;
			}
		}

		fmt.Println(totalWeight)

	}

}

File: minSpantree.c
#include <stdio.h>
#include <stdlib.h>

struct Vertex {
    int value;
    struct Vertex *parent;
};

struct Vertex* getRoot(struct Vertex *node) {
    return node->parent == node ? node : getRoot(node->parent);
}

struct Edge {
    struct Vertex *u;
    struct Vertex *v;
    int weight;
};

int edgeCompare(const void * a, const void * b) {
    return ( (*(struct Edge*)a).weight - (*(struct Edge*)b).weight );
}

int main(int argc, const char * argv[]) {
    
    int test;
    scanf("%d", &test);
    for (; test > 0; test--) {
        
        int numVertices;
        scanf("%d", &numVertices);
        int numEdges;
        scanf("%d", &numEdges);
        
        struct Vertex vertices[numVertices];
        struct Edge edges[numEdges];
        
        for (int i = 0; i < numVertices; i++) {
            vertices[i].value = i;
            vertices[i].parent = &vertices[i];
        }
        
        for (int i = 0; i < numEdges; i++) {
            int uValue;
            scanf("%d", &uValue);
            uValue--;
            int vValue;
            scanf("%d", &vValue);
            vValue--;
            edges[i].u = &vertices[uValue];
            edges[i].v = &vertices[vValue];
            scanf("%d", &edges[i].weight);
        }
        
        qsort(edges, numEdges, sizeof(edges[0]), edgeCompare);
        
        int totalWeight = 0;
        
        for (int i = 0; i < numEdges; i++) {
            if (getRoot(edges[i].u) != getRoot(edges[i].v)) {
                getRoot(edges[i].v)->parent = getRoot(edges[i].u);
                totalWeight += edges[i].weight;
            }
        }
        
        printf("%d\n", totalWeight);
        
    }
    
    return 0;
}
